staging/src/k8s.io/apimachinery/pkg/runtime/serializer/codec_factory.go

```diff
@@ -103,12 +109,16 @@ func NewCodecFactory(scheme *runtime.Scheme) CodecFactory {
 // newCodecFactory is a helper for testing that allows a different metafactory to be specified.
 func newCodecFactory(scheme *runtime.Scheme, serializers []serializerType) CodecFactory {
 	decoders := make([]runtime.Decoder, 0, len(serializers))
+	decodersStrict := make([]runtime.Decoder, 0, len(serializers))
 	var accepts []runtime.SerializerInfo
 	alreadyAccepted := make(map[string]struct{})
 
 	var legacySerializer runtime.Serializer
 	for _, d := range serializers {
 		decoders = append(decoders, d.Serializer)
+		if d.StrictSerializer != nil {
+			decodersStrict = append(decodersStrict, d.StrictSerializer)
```

Comment:
this could end up with an oddly asymmetric decoder if not all serializers provided a strict serializer, which doesn't seem good.

Reply:
i tough about this, but i don't know how we can ensure that all serializers are strict and the decoders and decodersStrict match.
the main reason for the nil check was that there is another protobuf serializer added to the list that is not setting a StrictSerializer.

on the other hand if one uses:
MyCodecFactory.UniversalStrictDecoder()
this guaranties that only the universalStrict list is going to be used.

Reply:
> the main reason for the nil check was that there is another protobuf serializer added to the list that is not setting a StrictSerializer.

that means that valid protobuf data with no unknown fields could fail strict decoding. that is unexpected.

Reply:
it seems to me that the protobuf does not have strict mode?
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf_extension.go#L33
V
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf.go#L154
V
https://github.com/gogo/protobuf/blob/master/proto/decode.go#L335

would it make sense to fallback to a non-strict unmarshall in the case of a nil:
```
if d.StrictSerializer != nil {
	decodersStrict = append(decodersStrict, d.StrictSerializer)
} else {
	decodersStrict = append(decodersStrict, d.Serializer)
}
```

as an extra action, it's also possible to add a StrictSerializer filed here being the same as the regular serializer:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf_extension.go#L39

Reply:
updated to follow the above strict vs non-strict.